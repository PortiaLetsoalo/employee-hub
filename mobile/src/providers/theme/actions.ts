import { createAction } from 'redux-actions';
import type { ThemeMode } from './utils/types';
import type { IThemeStateContext } from './contexts';
import type { IDrawerThemePalette, IThemePalette } from './utils/interface';

export enum FormsActionEnums {
  SetThemeModeAction = 'SET_THEME_MODE_ACTION',
  SetThemeAction = 'SET_THEME_ACTION',
  SetThemeColorsAction = 'SET_THEME_COLORS_ACTION',
  SetDrawerThemeColorsAction = 'SET_DRAWER_THEME_COLORS_ACTION',
}

export const setThemeAction = createAction<IThemeStateContext, any>(
  FormsActionEnums.SetThemeAction,
  (theme) => ({ theme })
);

export const setThemeModeAction = createAction<IThemeStateContext, ThemeMode>(
  FormsActionEnums.SetThemeModeAction,
  (themeMode) => ({ themeMode })
);

export const setThemeColorsAction = createAction<
  IThemeStateContext,
  IThemePalette
>(FormsActionEnums.SetThemeColorsAction, (themeColors) => ({ themeColors }));

export const setDrawerThemeColorsAction = createAction<
  IThemeStateContext,
  IDrawerThemePalette
>(FormsActionEnums.SetDrawerThemeColorsAction, (drawerThemeColors) => ({
  drawerThemeColors,
}));

/* NEW_ACTION_GOES_HERE */
