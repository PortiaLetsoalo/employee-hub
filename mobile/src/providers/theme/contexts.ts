import { createContext } from 'react';
//import type { IFlagsSetters, IFlagsState } from '../../utils/flags';
import type { ThemeMode } from './utils/types';
//import { FLAGS_INITIAL_STATE } from '../../utils/flags/flagsReducer';
import type {
  IDrawerThemePalette,
  IShaTheme,
  IThemePalette,
} from './utils/interface';

export type IFlagProgressFlags = '' /* NEW_IN_PROGRESS_FLAG_GOES_HERE */;
export type IFlagSucceededFlags = '' /* NEW_SUCCEEDED_FLAG_GOES_HERE */;
export type IFlagErrorFlags = '' /* NEW_ERROR_FLAG_GOES_HERE */;
export type IFlagActionedFlags = '' /* NEW_ACTIONED_FLAG_GOES_HERE */;

export interface IThemeStateContext
  // extends IFlagsState<
  //   IFlagProgressFlags,
  //   IFlagSucceededFlags,
  //   IFlagErrorFlags,
  //   IFlagActionedFlags
  // >
   {
  readonly theme?: IShaTheme;
  readonly themeColors?: IThemePalette;
  readonly themeMode?: ThemeMode;
  readonly drawerThemeColors?: IDrawerThemePalette;
}

export interface IThemeActionsContext
  // extends IFlagsSetters<
  //   IFlagProgressFlags,
  //   IFlagSucceededFlags,
  //   IFlagErrorFlags,
  //   IFlagActionedFlags
  // > 
  {
  getTheme: () => IThemePalette;
  getDrawerTheme: () => IDrawerThemePalette;
  //getThemeMode: () => any;
  setTheme: (theme: any) => void;
  setThemeMode: (mode: ThemeMode) => void;
}

export const FORMS_CONTEXT_INITIAL_STATE: IThemeStateContext = {
 // ...FLAGS_INITIAL_STATE,
  theme: {},
  themeColors: {},
  themeMode: 'light',
  drawerThemeColors: {},
};

export const FormsStateContext = createContext<IThemeStateContext>(
  FORMS_CONTEXT_INITIAL_STATE
);

export const FormsActionsContext = createContext<
  IThemeActionsContext | undefined
>(undefined);
