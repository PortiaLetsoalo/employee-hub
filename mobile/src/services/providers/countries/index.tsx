import React, {
  FC,
  PropsWithChildren,
  useContext,
  useMemo,
  useState,
} from "react";
import { CountriesService } from "../../generated/api";
import { AxiosError } from "axios";
import {
  ICountriesErrorResponse,
  ICountiriesStateContext,
  countriesState,
  countriesActions,
} from "./context";

const CountriesProvider: FC<PropsWithChildren<any>> = ({ children }) => {
  const [countries, setCountries] = useState<ICountiriesStateContext["countries"]>();

  const getCountries = () => {
    setCountries({ state: "loading" });
    CountriesService.getApiServicesAppCountriesGetAllCountries()
      .then((res) => {
        setCountries({
          state: "success",
          value: res["result"],
        });
      })
      .catch((error: AxiosError<ICountriesErrorResponse>) => {
        setCountries({ state: "error", error: error?.response?.data });
        console.log("UpdatePersonError", error);
      });
  };


  const memoizedState = useMemo(() => {
    return { countries };
  }, [countries]);

  const memoizedActions = useMemo(() => {
    return {
      getCountries,
    };
  }, [getCountries]);

  return (
    <countriesState.Provider value={memoizedState}>
      <countriesActions.Provider value={memoizedActions}>
        {children}
      </countriesActions.Provider>
    </countriesState.Provider>
  );
};

const useCountriestate = () => {
  const context = useContext(countriesState);
  return context;
};

const useCountriesActions = () => {
  const context = useContext(countriesActions);
  return context;
};

const useCountries = () => {
  return { ...useCountriestate(), ...useCountriesActions() };
};

export { CountriesProvider, useCountries, useCountriestate, useCountriesActions };
