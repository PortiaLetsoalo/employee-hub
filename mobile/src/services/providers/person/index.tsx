import React, {
  FC,
  PropsWithChildren,
  useContext,
  useMemo,
  useState,
} from "react";
import { PersonCreateDto, PersonService } from "../../generated/api";
import { AxiosError } from "axios";
import {
  IPersonErrorResponse,
  IPersonStateContext,
  PersonActions,
  PersonState,
} from "./context";

const PersonProvider: FC<PropsWithChildren<any>> = ({ children }) => {
  const [person, setPerson] = useState<IPersonStateContext["person"]>();
  const [persons, setPersons] = useState<IPersonStateContext["persons"]>();

  const getPersonById = (id: any) => {
    setPerson({ state: "loading" });
    PersonService.getApiServicesAppPersonGet(id)
      .then((res) => {
        setPerson({ state: "success", value: res["result"] });
      })
      .catch((error: AxiosError<IPersonErrorResponse>) => {
        setPerson({ state: "error", error: error?.response?.data });
      });
  };

 const createPerson = (payload: PersonCreateDto) : Promise<PersonCreateDto> =>
    new Promise((resolve, reject) => {
      setPerson({ state: "loading" });
    PersonService.postApiServicesAppPersonCreateAsyc(payload)
      .then((res) => {
        setPerson({ state: "success", value: res["result"] });
        resolve(res["result"] )
      }).catch((error: AxiosError<IPersonErrorResponse>) => {
        setPerson({ state: "error", error: error?.response?.data });
        reject(error?.response?.data)
      });
     });
  
  const updatePerson = (payload: PersonCreateDto) : Promise<PersonCreateDto> =>
    new Promise((resolve, reject) => {
      setPerson({ state: "loading" });
    PersonService.putApiServicesAppPersonUpdate(payload)
      .then((res) => {
        setPerson({ state: "success", value: res["result"] });
        resolve(res["result"] )
      }).catch((error: AxiosError<IPersonErrorResponse>) => {
        setPerson({ state: "error", error: error?.response?.data });
        reject(error?.response?.data)
      });
     });
  

  const getAllPerson = () => {
    setPerson({ state: "loading" });
    PersonService.getApiServicesAppPersonGetAll()
      .then((res) => {
        setPersons({
          state: "success",
          value:res['result']
        });
      })
      .catch((error: AxiosError<IPersonErrorResponse>) => {
        setPersons({ state: "error", error: error?.response?.data });
      });
  };



  const memoizedState = useMemo(() => {
    return { person ,persons};
  }, [person,persons]);

  const memoizedActions = useMemo(() => {
    return {
      getPersonById,
      createPerson,
      updatePerson,
      getAllPerson
    };
  }, [createPerson, getPersonById, updatePerson,getAllPerson]);

  return (
    <PersonState.Provider value={memoizedState}>
      <PersonActions.Provider value={memoizedActions}>
        {children}
      </PersonActions.Provider>
    </PersonState.Provider>
  );
};

const usePersonState = () => {
  const context = useContext(PersonState);
  return context;
};

const usePersonActions = () => {
  const context = useContext(PersonActions);
  return context;
};

const usePerson = () => {
  return { ...usePersonState(), ...usePersonActions() };
};

export { PersonProvider, usePerson, usePersonState, usePersonActions };
