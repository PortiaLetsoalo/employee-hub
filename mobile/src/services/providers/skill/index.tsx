import React, {
  FC,
  PropsWithChildren,
  useContext,
  useMemo,
  useState,
} from "react";
import { SkillDto, SkillService } from "../../generated/api";
import { AxiosError } from "axios";
import {
  ISkillErrorResponse,
  ISkillStateContext,
  SkillActions,
  SkillState,
} from "./context";

const SkillProvider: FC<PropsWithChildren<any>> = ({ children }) => {
  const [skill] = useState<ISkillStateContext["skill"]>();
  const [skills, setSkills] = useState<ISkillStateContext["skills"]>();
  const [personSkills, setPersonSkills] = useState<ISkillStateContext["personSkills"]>();

  const getSkillsByPersonId = (id: any) => {
    setPersonSkills({ state: "loading" });
    SkillService.getApiServicesAppSkillGetAllByPersonId(id)
      .then((res) => {
        setPersonSkills({ state: "success", value: res["result"] });
      })
      .catch((error: AxiosError<ISkillErrorResponse>) => {
        setPersonSkills({ state: "error", error: error?.response?.data });
      });
  };

  const createSkillSet = (payload: SkillDto[]) => {
    setSkills({ state: "loading" });
    SkillService.postApiServicesAppSkillCreateSkillSet(payload)
      .then((res) => {
        setSkills({
          state: "success",
          value: res["result"],
        });
      })
      .catch((error: AxiosError<ISkillErrorResponse>) => {
        setSkills({ state: "error", error: error?.response?.data });
      });
  };

  const updateSkillSet = (payload: SkillDto[]) => {
    setSkills({ state: "loading" });
    SkillService.putApiServicesAppSkillUpdateSkillSet(payload)
      .then((res) => {
        setSkills({
          state: "success",
          value:res['result']
        });
      })
      .catch((error: AxiosError<ISkillErrorResponse>) => {
        setSkills({ state: "error", error: error?.response?.data });
      });
  };

  const deleteSkill = (id: any) => {
     setSkills({ state: "loading" });
     SkillService.deleteApiServicesAppSkillDelete(id).then((res) => {
        setSkills({
          state: "success",
          value:res['result']
        });
      })
  }

  const getAllSkills = () => {
    setSkills({ state: "loading" });
    SkillService.getApiServicesAppSkillGetAll()
      .then((res) => {
        setSkills({
          state: "success",
          value:res['result']
        });
      })
      .catch((error: AxiosError<ISkillErrorResponse>) => {
        setSkills({ state: "error", error: error?.response?.data });
      });
  };

  const memoizedState = useMemo(() => {
    return { skill, skills,personSkills };
  }, [skill,skills,personSkills]);

  const memoizedActions = useMemo(() => {
    return {
      getSkillsByPersonId,
      createSkillSet,
      updateSkillSet,
      getAllSkills,
      deleteSkill
    };
  }, [getSkillsByPersonId, createSkillSet, updateSkillSet,getAllSkills,deleteSkill]);

  return (
    <SkillState.Provider value={memoizedState}>
      <SkillActions.Provider value={memoizedActions}>
        {children}
      </SkillActions.Provider>
    </SkillState.Provider>
  );
};

const useSkillState = () => {
  const context = useContext(SkillState);
  return context;
};

const useSkillActions = () => {
  const context = useContext(SkillActions);
  return context;
};

const useSkill = () => {
  return { ...useSkillState(), ...useSkillActions() };
};

export { SkillProvider, useSkill, useSkillState, useSkillActions };
