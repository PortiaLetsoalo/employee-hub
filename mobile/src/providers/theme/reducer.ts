//import flagsReducer from '../../utils/flags/flagsReducer';
import { FormsActionEnums } from './actions';
import type { IThemeStateContext } from './contexts';

export function themeReducer(
  incomingState: IThemeStateContext,
  action: ReduxActions.Action<IThemeStateContext>
): IThemeStateContext {
//const state = flagsReducer(action, incomingState);
const state:any = ''
  const { type, payload } = action;

  switch (type) {
    case FormsActionEnums.SetThemeAction:
    case FormsActionEnums.SetThemeModeAction:
    case FormsActionEnums.SetThemeColorsAction:
    case FormsActionEnums.SetDrawerThemeColorsAction:
      /* NEW_ACTION_ENUM_GOES_HERE */
      return {
        ...state,
        ...payload,
      };

    default: {
      return state;
    }
  }
}
