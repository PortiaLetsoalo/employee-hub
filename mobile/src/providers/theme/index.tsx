import React, {
  FC,
  PropsWithChildren,
  useContext,
  useEffect,
  useReducer,
} from 'react';
import type { ThemeMode } from './utils/types';
//import { getFlagSetters } from '../../utils/flags/flagsSetters';
import {
  setThemeModeAction,
  setThemeAction,
  setThemeColorsAction,
  setDrawerThemeColorsAction,
} from './actions';
import {
  FormsActionsContext,
  FormsStateContext,
  FORMS_CONTEXT_INITIAL_STATE,
} from './contexts';
import { themeReducer } from './reducer';
import type {
  IDrawerThemePalette,
  IShaThemeProviderProps,
  IThemePalette,
} from '../theme/utils/interface';

const ThemeProvider: FC<PropsWithChildren<IShaThemeProviderProps>> = ({
  themeMode,
  theme,
  children,
}) => {
  const [state, dispatch] = useReducer(
    themeReducer,
    FORMS_CONTEXT_INITIAL_STATE
  );

  useEffect(() => {
    updateThemes();
  }, [themeMode, theme]);

  const updateThemes = () => {
    setTheme(theme);
    setThemeMode(themeMode ?? 'light');
    setThemeColors(getTheme() ?? {});
    setDrawerThemeColors(getDrawerTheme() ?? {});
  };

  const getTheme = (): IThemePalette => {
    let activeTheme: IThemePalette | undefined = theme?.colors?.light;
    if (themeMode == 'dark') {
      activeTheme = theme?.colors?.dark;
    }
    return activeTheme ?? {};
  };

  const getDrawerTheme = (): IDrawerThemePalette => {
    let activeDrawerTheme: IDrawerThemePalette | undefined =
      theme?.colors?.drawerLight;
    if (theme?.sidebar == 'dark') {
      activeDrawerTheme = theme?.colors?.drawerDark;
    }
    return activeDrawerTheme ?? {};
  };

  const setTheme = (theme: any) => {
    dispatch(setThemeAction(theme));
  };
  const setThemeColors = (themeColors: IThemePalette) => {
    dispatch(setThemeColorsAction(themeColors));
  };

  const setDrawerThemeColors = (drawerThemeColors: IDrawerThemePalette) => {
    dispatch(setDrawerThemeColorsAction(drawerThemeColors));
  };

  const setThemeMode = (mode: ThemeMode) => {
    dispatch(setThemeModeAction(mode));
  };

  return (
    <FormsStateContext.Provider value={{ ...state, themeMode, theme }}>
      <FormsActionsContext.Provider
        value={{
         // ...getFlagSetters(dispatch),
          getTheme,
          getDrawerTheme,
          setTheme,
          setThemeMode,
        }}
      >
        {children}
      </FormsActionsContext.Provider>
    </FormsStateContext.Provider>
  );
};

function useThemeState() {
  const context = useContext(FormsStateContext);

  if (context === undefined) {
    // throw new Error('useThemeState must be used within a FormsProvider');
  }

  return context;
}

function useThemeActions() {
  const context = useContext(FormsActionsContext);

  if (context === undefined) {
    // throw new Error('useThemeActions must be used within a FormsProvider');
  }

  return context;
}

function useTheme() {
  return { ...useThemeState(), ...useThemeActions() };
}

export { ThemeProvider, useThemeState, useThemeActions, useTheme };
