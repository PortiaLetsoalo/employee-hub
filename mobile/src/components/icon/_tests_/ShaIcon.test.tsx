import React from 'react';
import {create, act} from 'react-test-renderer';
import {Icon} from '@rneui/themed';
import ShaIcon from '../shaIcon';
import {useTheme} from '../../../providers/theme';

jest.mock('@rneui/themed', () => ({
  Icon: 'Icon',
}));

jest.mock('../../../providers/theme', () => ({
  useTheme: jest.fn(),
}));

describe('ShaIcon Component', () => {
  let tree;
  const mockTheme = {
    themeColors: {
      iconColor: 'blue',
    },
  };

  beforeEach(() => {
    (useTheme as jest.Mock).mockReturnValue(mockTheme);

    const defaultProps = {
      name: 'home',
      type: 'font-awesome',
      size: 24,
      color: 'red',
      onPress: jest.fn(),
      containerStyle: {padding: 10},
      iconStyle: {margin: 5},
    };

    tree = create(<ShaIcon {...defaultProps} />);
  });

  it('matches snapshot', () => {
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('is defined', () => {
    expect(
      <ShaIcon name={'home'} type={'antdesign'} size={14} />,
    ).toBeDefined();
  });

  it('renders with correct props', () => {
    const instance = tree.root;
    expect(instance.findByType(Icon).props.name).toEqual('home');
    expect(instance.findByType(Icon).props.type).toEqual('font-awesome');
    expect(instance.findByType(Icon).props.size).toEqual(24);
    expect(instance.findByType(Icon).props.color).toEqual('red');
    expect(instance.findByType(Icon).props.containerStyle).toEqual({
      padding: 10,
    });
    expect(instance.findByType(Icon).props.iconStyle).toEqual({margin: 5});
  });

  it('handles onPress correctly', () => {
    const instance = tree.root;
    const onPressMock = instance.props.onPress;

    act(() => {
      instance.findByType(Icon).props.onPress();
    });

    expect(onPressMock).toHaveBeenCalled();
  });

  it('updates props correctly', () => {
    act(() => {
      tree.update(
        <ShaIcon
          name="user"
          type="material"
          size={30}
          color="green"
          containerStyle={{margin: 20}}
          iconStyle={{padding: 10}}
        />,
      );
    });

    const instance = tree.root;
    expect(instance.findByType(Icon).props.name).toEqual('user');
    expect(instance.findByType(Icon).props.type).toEqual('material');
    expect(instance.findByType(Icon).props.size).toEqual(30);
    expect(instance.findByType(Icon).props.color).toEqual('green');
    expect(instance.findByType(Icon).props.containerStyle).toEqual({
      margin: 20,
    });
    expect(instance.findByType(Icon).props.iconStyle).toEqual({padding: 10});
  });
});
