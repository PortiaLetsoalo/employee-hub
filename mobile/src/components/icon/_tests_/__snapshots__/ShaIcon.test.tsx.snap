// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShaIcon Component matches snapshot 1`] = `
<Icon
  color="red"
  containerStyle={
    Object {
      "padding": 10,
    }
  }
  iconStyle={
    Object {
      "margin": 5,
    }
  }
  name="home"
  onPress={[MockFunction]}
  size={24}
  type="font-awesome"
/>
`;
